### DISCUSSION TITLE
What is your custom Start GCODE?

### DISCUSSION THREAD

**Discussion by mrarzt** [mriscoc/Ender3V2S1#24] [General]
I´m using this but actual after i resetted my printer it seems not to work good. No print starts... :-/

```g-code
M190 S{first_layer_bed_temperature[0]} ; set final bed temp
G28 ;Home
G29 ;ABL
M500 ;save bed level
M420 S1 ;enable ABL to get sure
;*** Start Preheating ***
G1 X8.0 Y15 Z30.0 F5000.0 ;No blob on Heatbed
M109 S{first_layer_temperature[0]} ; wait for nozzle temp to stabilize
;*** End Preheating ***
G92 E0 ;Reset Extruder
G1 Z2.0 F3000 ;Move Z Axis up
G1 X12.0 Y60 Z0.28 F5000.0 ;Move to start position
G1 X12.0 Y190.0 Z0.28 F1500.0 E15 ;Draw the first line
G1 X14.0 Y190.0 Z0.28 F5000.0 ;Move to side a little
G1 X14.0 Y40 Z0.28 F1500.0 E30 ;Draw the second line
G92 E0 ;Reset Extruder
G1 Z2.0 F3000 ;Move Z Axis up
```

**Comment by mriscoc** [comment]
Your bed is heating while the mesh is taken, that is not good, is better to do the mesh when the bed is thermally stable.

**Comment by mrarzt** [comment]
Mmmmh... but this code is doing this? M190 is waiting for the bed.

Which code did you use?

**Reply by mriscoc** [reply]
M190 waits to the temperature in the thermistor not really for thermal stabilization.
I don't make the mesh on each print. I only put a M420 S1 Z2 after the last G28.

**Reply by mriscoc** [reply]
https://github.com/MarlinFirmware/Marlin/pull/22005

**Reply by mrarzt** [reply]
Interesting. Can you copy paste please your start and end code?

**Comment by mriscoc** [comment]
## For Cura
Start g-code:
```g-code
M201 X500.00 Y500.00 Z100.00 E5000.00 ;Setup machine max acceleration
M203 X500.00 Y500.00 Z10.00 E50.00 ;Setup machine max feedrate
M204 P500.00 R1000.00 T500.00 ;Setup Print/Retract/Travel acceleration
M205 X8.00 Y8.00 Z0.40 E5.00 ;Setup Jerk
M220 S100 ;Reset Feedrate
M221 S100 ;Reset Flowrate

; Ender 3 Custom Start G-code
G92 E0 ; Reset Extruder
G28 ; Home all axes
M420 S1 Z2; Use Level Mesh up to 2mm;
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position
G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line
G92 E0 ; Reset Extruder
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish
```

End g-code
```g-code
G91 ;Relative positioning
G1 E-2 F2700 ;Retract a bit
G1 E-2 Z0.2 F2400 ;Retract and raise Z
G1 X5 Y5 F3000 ;Wipe out
G1 Z10 ;Raise Z more
G90 ;Absolute positioning

G1 X0 Y{machine_depth} ;Present print
M106 S0 ;Turn-off fan
M104 S0 ;Turn-off hotend
M140 S0 ;Turn-off bed

M84 X Y E ;Disable all steppers but Z
```

**Reply by fabiomaticus** [reply]
Hi, what is {machine_depth}? in end g-code?

**Reply by BrianSPC** [reply]
Can this Start and End code be used in Prusa/SuperSlicer ?

**Reply by mriscoc** [reply]
For Prusa/SuperSlicer you must make small changes.

**Reply by leydonhd** [reply]
in Prusa where and what changes must be done?

**Comment by john-helyar** [comment]
Hi this is my Gcode im new to the world of 3D Printing have I set this up correctly so that it warms up and then does an ABL before each print?

``` Ender 3 Custom Start G-code
G92 E0 ; Reset Extruder
M140 S{material_bed_temperature_layer_0} ; start preheating the bed
M104 S{material_print_temperature_layer_0} ﻿T0 ; start preheating hotend
G28 ; home
M190 S{material_bed_temperature_layer_0} ; heat to Cura Bed setting 
M109 S{material_print_temperature_layer_0} ﻿T0 ; heat to Cura Hotend
G1 E-15 F2700 ;Retract a bit
G29 ; Level Print Bed
M420 S1 ; Use Last Mesh
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position
G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line
G92 E0 ; Reset Extruder
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish```

**Reply by mriscoc** [reply]
You can add a C108 after the G29 to avoid pressing the continue button.

**Comment by john-helyar** [comment]
Hi there i added the C108 on the line after G29 it stays on screen showing the Mesh its made is there away to auto save this it does carry on but wondered if there was a way to save and clear the screen back to the main screen ?

**Reply by mriscoc** [reply]
To save settings to the EEPROM: `M500`
Close Mesh Viewer: `C108`
If you are printing the screen go back to printing screen.

**Comment by john-helyar** [comment]
ok so this is my gcode now?

```; Ender 3 Custom Start G-code
G92 E0 ; Reset Extruder
M140 S{material_bed_temperature_layer_0} ; start preheating the bed
M104 S{material_print_temperature_layer_0} ﻿T0 ; start preheating hotend
G28 ; home
G1 E-15 F2700 ;Retract a bit
M190 S{material_bed_temperature_layer_0} ; heat to Cura Bed setting 
M109 S{material_print_temperature_layer_0} ﻿T0 ; heat to Cura Hotend
G29 ; Level Print Bed
M500 ; Save to EEPROM
C108 ; Close Mesh Viewer
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position
G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line
G92 E0 ; Reset Extruder
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish
```
Would this be correct?

**Comment by fabiomaticus** [comment]
In prusa slicer, what is the start gcode and end gcode?

**Reply by mriscoc** [reply]
https://github.com/mriscoc/Ender3V2S1/wiki/Slicer-G-code-Scripts

**Comment by fabiomaticus** [comment]
I copied the prusa code, but I have this error:
"" failed due to invalid custom G-code sections:
start_gcode
Parsing error at line 18: Not a variable name
M104 S{first_layer_temperature[initial_extruder]+extruder_temperature_offset[initial_extruder]} ; set final nozzle temp
                                                 ^
        !!!!! Failed to process the custom G-code template ...
and
        !!!!! End of an error report for the custom G-code template ...
for all macro processing errors.""

**Reply by mriscoc** [reply]
Try replacing `initial_extruder` by 0.

**Reply by fabiomaticus** [reply]
I try this, but i have same isse:

`M501; Read values from EEPROM
M201 X500.00 Y500.00 Z100.00 E5000.00 ;Setup machine max acceleration
M203 X500.00 Y500.00 Z20.00 E50.00 ;Setup machine max feedrate
M204 P500.00 R1000.00 T500.00 ;Setup Print/Retract/Travel acceleration
M205 X8.00 Y8.00 Z0.40 E5.00 ;Setup Jerk
M220 S100 ;Reset Feedrate
M221 S100 ;Reset Flowrate

G90 ; use absolute coordinates
M83 ; extruder relative mode
M104 S120 ; set temporary nozzle temp to prevent oozing during homing and auto bed leveling
M140 S[first_layer_bed_temperature] ; set final bed temp
G4 S10 ; allow partial nozzle warmup
G28 ; home all axis
M420 S1 Z2 ; Use mesh level upto 2 mm
G1 Z50 F240
G1 X2 Y10 F3000
M104 S{first_layer_temperature[0]+extruder_temperature_offset[0]} ; set final nozzle temp
M190 S[first_layer_bed_temperature] ; wait for bed temp to stabilize
M109 S{first_layer_temperature[0]+extruder_temperature_offset[0]} ; wait for nozzle temp to stabilize
G1 Z0.28 F240
G92 E0
G1 Y140 E10 F1500 ; prime the nozzle
G1 X2.3 F5000
G92 E0
G1 Y10 E10 F1200 ; prime the nozzle
G92 E0`

**Comment by Zuendapp** [comment]
Hi :),

by the way, its a very nice firmware, really!!
Is the Start-Code another when i use the UBL Firmware for my Ender 3V2 (422) with a CR Touch?

I used so far following start-code after G28 
 
  .....
  G28 ; Home all axes
  G29 L0 ; load mesh from slot 0
  G29 A ; activate unified bed leveling
  G29 J2 ; tilt mesh with 4-pt check
  .....

..or should I use your suggestion "M420 S1 Z2; Use Level Mesh up to 2mm;" instead my UBL-Code?

Thank you very much!

BR

**Reply by mriscoc** [reply]
Hi, seems ok for a UBL version.

**Reply by Zuendapp** [reply]
Thank you!!
